### PR TITLE
[ml/tune] Expose new checkpoint interface to users

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This PR exposes the new checkpoint interface, implemented in #22691, to end users. It does this by replacing the old external facing `TrialCheckpoint` class with a merged class that supports the old TrialCheckpoint API (upload, download, save) as well as the new Checkpoint API.

With this PR, users can use the new Checkpoint interface for downstream processing of their Ray Tune results. In a follow-up PR, the new Checkpoint interface will be used internally within Ray Tune and Train for bookkeeping, however, that is not required to unblock the Ray ML use case.

## Related issue number

Depends on #22691

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xwjiang2010** [comment]
Looking good. So the idea is that both `_TrialCheckpoint` and `TrialCheckpoint` are being deprecated. After a few cycles, we will converge to using just Checkpoint.

**Comment by krfricke** [comment]
That's correct!

Cleaned up the PR after merging #22691, review should be straightforward

**Review by Yard1 — APPROVED** [review]
LGTM

File: python/ray/tune/cloud.py [hunk 1]
```diff
@@ -295,3 +317,107 @@ def save(self, path: Optional[str] = None, force_download: bool = False):
...
+        )
+        return _TrialCheckpoint.upload(self, cloud_path, local_path, clean_before)
+
+    def save(self, path: Optional[str] = None, force_download: bool = False):
+        warnings.warn(
+            "`checkpoint.save()` is deprecated and will be removed in "
+            "the future. Please use `checkpoint.to_directory()` or"
+            "`checkpoint.to_uri()` instead.",
+            DeprecationWarning,
+        )
```
Comment (richardliaw) [inline python/ray/tune/cloud.py:422]: Can you protect this with a 'log_once'?

Comment (krfricke) [reply python/ray/tune/cloud.py]: Sure! Done

**Comment by richardliaw** [comment]
BTW this PR deprecates things and doesnt expose any new interfaces right?

**Comment by ericl** [comment]
LGTM. We should also make sure Tuner only exposes the new Checkpoint interface.

**Review by xwjiang2010 — APPROVED** [review]

**Comment by krfricke** [comment]
@richardliaw it deprecates the old external facing TrialCheckpoint class but exposes the new Checkpoint class. This is done by merging them and throwing warnings when the old TrialCheckpoint API is used. In the release after the next we'll raise an error and in the one after that we'll remove the old TrialCheckpoint class completely.

**Comment by krfricke** [comment]
Good point, I've added a simple switch in the `ExperimentAnalysis` that the `Tuner` can use to only return `Checkpoint` objects. cc @xwjiang2010

**Comment by xwjiang2010** [comment]
> LGTM. We should also make sure Tuner only exposes the new Checkpoint interface.

Ah got it. Thanks Kai!

**Pull request merged**